calculator.go

```diff
@@ -1,13 +1,96 @@
 // Package calculator provides a library for simple calculations in Go.
 package calculator
 
+import (
+	"go/token"
+	"go/types"
+	"math"
+	"strconv"
+)
+
+const divideByZeroMessage = "Cannot divide by zero"
+const negativeNumber = "No negative number allowed"
+
+// ErrorString is a trivial implementation of error.
+type ErrorString struct {
+	s string
+}
+
+func (e *ErrorString) Error() string {
+	return e.s
+}
+
+func extractFirstItem(numbers []float64) (float64, []float64) {
+	return numbers[0], numbers[1:]
+}
+
 // Add takes two numbers and returns the result of adding them together.
-func Add(a, b float64) float64 {
-return a + b
+func Add(numbers ...float64) float64 {
+	var total float64 = 0
+	for i := 0; i < len(numbers); i++ {
+		total += numbers[i]
+	}
+	return total
 }
 
 // Subtract takes two numbers and returns the result of subtracting the second
 // from the first.
-func Subtract(a, b float64) float64 {
-	return b - a
+func Subtract(numbers []float64) float64 {
+	firstValue, values := extractFirstItem(numbers)
+	for i := 0; i < len(values); i++ {
+		firstValue -= values[i]
+	}
+	return firstValue
+}
+
+// Multiply takes two numbers and returns the result of multiplication them together.
+func Multiply(numbers []float64) float64 {
+	firstValue, values := extractFirstItem(numbers)
+	for i := 0; i < len(values); i++ {
+		firstValue *= values[i]
+	}
+	return firstValue
+}
+
+// Divide takes two numbers and returns the result of divion and an error message
+func Divide(numbers []float64) (float64, error) {
+	firstValue, values := extractFirstItem(numbers)
+	for i := 0; i < len(values); i++ {
+		if values[i] == 0 {
+			return 0, &ErrorString{divideByZeroMessage}
+		}
+		firstValue /= values[i]
```

Comment:
A problem with modifying `firstValue` like this is that it's no longer the first value! We're effectively creating a new variable which holds the current _result_, which will eventually be returned. So `result` might be a better name for it.